OpenRA.Game/Activities/Activity.cs

```diff
@@ -59,17 +56,49 @@ public Activity TickOuter(Actor self)
 				State = ActivityState.Active;
 			}
 
-			var ret = Tick(self);
-			if (ret != this)
+			// Only run the parent tick when the child is done.
+			// We must always let the child finish on its own before continuing.
+			if (ChildHasPriority)
+			{
+				lastRun = TickChild(self) && (finishing || Tick(self));
+				finishing |= lastRun;
+			}
+
+			// The parent determines whether the child gets a chance at ticking.
+			else
+				lastRun = Tick(self);
+
+			// Avoid a single tick delay if the childactivity was just queued.
+			if (ChildActivity != null && ChildActivity.State == ActivityState.Queued)
+				lastRun = TickChild(self);
+
+			if (lastRun)
 			{
 				State = ActivityState.Done;
 				OnLastRun(self);
+				return NextActivity;
 			}
 
-			return ret;
+			return this;
 		}
 
-		public abstract Activity Tick(Actor self);
+		protected bool TickChild(Actor self)
+		{
+			ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+			if (ChildActivity != null)
+				return false;
+
+			return true;
+		}
+
+		/// <summary>
```

Comment:
It would be good if we could add some more detail here.
Maybe something like:
```xml
<summary>
Called every tick to run activity logic. Returns false if the activity should
remain active, or true if it is complete. Cancelled activities must ensure they
return the actor to a consistent state before returning true.

Child activities can be queued using QueueChild, and these will be ticked
instead of the parent while they are active. Activities that need to run logic
in parallel with child activities should set ChildHasPriority to false and
manually call TickChildren.

Queuing one or more child activities and returning true is valid, and causes
the activity to be completed immediately (without ticking again) once the
children have completed.
</summary>
```
?

Reply:
Done.